### PR TITLE
Add and populate leaf_id in boosted_trees.DebugOutput

### PR DESCRIPTION
**Pull request by bmc2-stripe**
Adds leaf_id field to boosted_trees.DebugOutput, and populates it in BoostedTreesExampleDebugOutputsOp.

Motivation:
*Leaf ids are quite useful in circumstances where we're using boosted trees to generate features for a subsequent model, eg in https://research.fb.com/publications/practical-lessons-from-predicting-clicks-on-ads-at-facebook/.
*This PR also (partially) resolves a TODO.

Tested:
Updated unit tests to validate correct leaf_ids are returned.

### CODE REVIEW

**Comment by bmc2-stripe** [comment]
@gbaned @jaingaurav 
Hi there! Could you please take a look, or re-assign? Thanks!
-Brendan

**Comment by mihaimaruseac** [comment]
Adding @tanzhenyu instead as I'm not qualified to review this

**Comment by bmc2-stripe** [comment]
Hi there @tanzhenyu -- could you please take a look, or re-assign? Thanks!

**Comment by bmc2-stripe** [comment]
Friendly ping @tanzhenyu

**Comment by gbaned** [comment]
@tanzhenyu Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@tanzhenyu Can you please review this PR ? Thanks!

**Comment by gbaned** [comment]
@tanzhenyu Can you please review this PR ? Thanks!

**Comment by bmc2-stripe** [comment]
ping @gbaned @mihaimaruseac @ebrevdo @tanzhenyu 

This has been pending for more than 7 months now.

**Comment by mihaimaruseac** [comment]
As mentioned, I am not qualified to review this. @tanzhenyu can you take a look or reassign please?

**Comment by tanzhenyu** [comment]
I don't have the bandwidth to review this, un-assigning myself.

File: tensorflow/core/kernels/boosted_trees/boosted_trees.proto [hunk 1]
```diff
@@ -162,9 +162,7 @@ message TreeEnsemble {
 }
 
 // DebugOutput contains outputs useful for debugging/model interpretation, at
-// the individual example-level. Debug outputs that are available to the user
-// are: 1) Directional feature contributions (DFCs) 2) Node IDs for ensemble
-// prediction path 3) Leaf node IDs.
+// the individual example-level.
```
Comment (nataliaponomareva) [inline tensorflow/core/kernels/boosted_trees/boosted_trees.proto:165]: i would suggest to keep this description 
 // Debug outputs that are available to the user
// are: 1) Directional feature contributions (DFCs) 2) Node IDs for ensemble
// prediction path 3) Leaf node IDs. 
And add that you also now have leaf ids available

**Review by nataliaponomareva — APPROVED** [review]

**Comment by gbaned** [comment]
@bmc2-stripe  Can you please take a look on the above comment from @nataliaponomareva. Thanks!

**Review by nataliaponomareva — APPROVED** [review]

**Review by mihaimaruseac — APPROVED** [review]

**Pull request merged**